### PR TITLE
detect: Add byte_math detector

### PR DESCRIPTION
**Pull request by jlucovsky**
Continuation of #4502 
This PR adds `byte_math` to the suite of `byte` functions with a strict implementation based on the Snort documents.

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: [2311](https://redmine.openinfosecfoundation.org/issues/2311)

Describe changes:
- Rebased

### CODE REVIEW

File: src/detect-byte.c [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ * \param arg The name of the variable being sought
+ * \param s The signature to check for the variable
+ * \param index When found, the value of the slot within the byte vars
+ *
+ * \retval true A match for the variable was found.
+ * \retval false
+ */
+bool DetectByteRetrieveSMVar(const char *arg, const Signature *s, DetectByteIndexType *index)
+{
+        SigMatch *bed_sm = DetectByteExtractRetrieveSMVar(arg, s);
```
Comment (victorjulien) [inline src/detect-byte.c:42]: indent looks off here

File: src/detect-byte.c [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ * \author Jeff Lucovsky <jeff@lucovsky.org>
+ */
+
+#include "suricata-common.h"
+#include "detect-byte.h"
+#include "detect-byte-extract.h"
+#include "detect-bytemath.h"
+
+/**
+ * \internal
```
Comment (victorjulien) [inline src/detect-byte.c:30]: we usually use `\internal` for `static` functions

File: src/detect-bytemath.c [hunk 3]
```diff
@@ -0,0 +1,1364 @@
...
+            SCLogDebug("error extracting %d bytes of numeric data: %d",
+                       data->nbytes, extbytes);
+            return 0;
+        }
+    }
+    ptr += extbytes;
+    det_ctx->buffer_offset = ptr - payload;
+
+    switch (data->oper) {
+    	case DETECT_BYTEMATH_OPERATOR_NONE:
```
Comment (victorjulien) [inline src/detect-bytemath.c:225]: `case` should indent 4 spaces from `switch`

File: src/detect-bytemath.c [hunk 4]
```diff
@@ -0,0 +1,1364 @@
...
+        int endianness = (endian == DETECT_BYTEMATH_ENDIAN_BIG) ?
+                          BYTE_BIG_ENDIAN : BYTE_LITTLE_ENDIAN;
+        extbytes = ByteExtractUint64(&val, endianness, data->nbytes, ptr);
+        if (extbytes != data->nbytes) {
+            SCLogDebug("error extracting %d bytes of numeric data: %d",
+                       data->nbytes, extbytes);
+            return 0;
+        }
+    }
+    ptr += extbytes;
```
Comment (victorjulien) [inline src/detect-bytemath.c:221]: should we first validate that adding `extbytes` doesn't overrun `ptr`?

File: src/detect-bytemath.c [hunk 5]
```diff
@@ -0,0 +1,1364 @@
...
+ * \brief Used to parse byte_math arg.
+ *
+ * \param arg The argument to parse.
+ * \param rvalue May be NULL. When non-null, will contain the variable
+ *              name of rvalue (iff rvalue is not a scalar value)
+ *
+ * \retval bmd On success an instance containing the parsed data.
+ *            On failure, NULL.
+ */
+static inline DetectByteMathData *DetectByteMathParse(const char *arg, char **rvalue)
```
Comment (victorjulien) [inline src/detect-bytemath.c:269]: don't think it's needed to make this `inline`

File: src/detect-bytemath.c [hunk 6]
```diff
@@ -0,0 +1,1364 @@
...
+        SCLogError(SC_ERR_INVALID_SIGNATURE, "byte_math invalid offset "
+                   "value \"%s\"", tmp_str);
+        goto error;
+    }
+
+    /* operator */
+    res = pcre_copy_substring((char *)arg, ov,
+                             MAX_SUBSTRINGS, OPER_VAL, tmp_str, sizeof(tmp_str));
+    if (res < 0) {
+        SCLogError(SC_ERR_PCRE_GET_SUBSTRING, "pcre_copy_substring failed "
```
Comment (victorjulien) [inline src/detect-bytemath.c:326]: nit: think we have `SC_ERR_PCRE_COPY_SUBSTRING` now

File: src/detect-bytemath.c [hunk 7]
```diff
@@ -0,0 +1,1364 @@
...
+                           "when used with \"dce\"; %d is not valid", bmd->nbytes);
+                goto error;
+                break;
+        }
+    }
+
+    switch (bmd->oper) {
+    	case DETECT_BYTEMATH_OPERATOR_LSHIFT:
+    	case DETECT_BYTEMATH_OPERATOR_RSHIFT:
+            /* nbytes has already been validated to be in [1, 10] */
```
Comment (victorjulien) [inline src/detect-bytemath.c:547]: whats this `[1, 10]`?

File: src/detect-bytemath.c [hunk 8]
```diff
@@ -0,0 +1,1364 @@
...
+            } else {
+                sm_list = SigMatchListSMBelongsTo(s, prev_pm);
+                if (sm_list < 0)
+                    goto error;
+            }
+        } else {
+            sm_list = DETECT_SM_LIST_PMATCH;
+        }
+
+        if (DetectSignatureSetAppProto(s, ALPROTO_DCERPC) < 0)
```
Comment (victorjulien) [inline src/detect-bytemath.c:626]: wonder if this is correct for a case where dcerpc is over smb. Do you have a test case for that?

File: src/detect-bytemath.c [hunk 9]
```diff
@@ -0,0 +1,1364 @@
...
+        }
+        data->rvalue = index;
+        data->flags |= DETECT_BYTEMATH_RVALUE_VAR;
+        SCFree(rvalue);
+        rvalue = NULL;
+    }
+
+    SigMatch *prev_bmd_sm = DetectGetLastSMByListId(s, sm_list,
+            DETECT_BYTEMATH, -1);
+    if (prev_bmd_sm == NULL)
```
Comment (victorjulien) [inline src/detect-bytemath.c:678]: I would find this code more readable with curly braces

File: src/detect-bytemath.c [hunk 10]
```diff
@@ -0,0 +1,1364 @@
...
+/*************************************Unittests********************************/
+#ifdef UNITTESTS
+
+#include "app-layer-parser.h"
+
+static int DetectByteMathParseTest01(void)
+{
+    int result;
+
+    DetectByteMathData *bmd = DetectByteMathParse("bytes 4, offset 2, oper +,"
```
Comment (victorjulien) [inline src/detect-bytemath.c:770]: new tests should be written using `FAIL_IF`/`PASS` macros w/o any conditional statements.

File: src/detect-bytemath.h [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+ */
+typedef struct DetectByteMathData_ {
+    /* local id used by other keywords in the sig to reference this */
+    uint8_t local_id;
+    uint8_t nbytes;
+    uint16_t offset;
+
+    uint32_t rvalue;
+
+    const char *result; /* consumed */
```
Comment (victorjulien) [inline src/detect-bytemath.h:59]: what is this used for?

File: src/detect-bytemath.c (hunk 7 above)
Comment (jlucovsky) [reply src/detect-bytemath.c:547]: I'll update the comment. The value has already been validated to be in the range `[1-10]`.

File: src/detect-bytemath.h (hunk 11 above)
Comment (jlucovsky) [reply src/detect-bytemath.h:59]: The `result` variable, if present. I'll add a brief comment.

File: src/detect-bytemath.c (hunk 8 above)
Comment (jlucovsky) [reply src/detect-bytemath.c:626]: I do not have a test case ... the pattern followed is the same as that in `byte_extract`

**Comment by jlucovsky** [comment]
Continued in #4823

File: src/detect-bytemath.c (hunk 8 above)
Comment (victorjulien) [reply src/detect-bytemath.c:626]: think we can create a SV test based on one of the smb/dce pcaps

**Pull request closed without merging**